src/nanoemoji/util.py

```diff
@@ -85,30 +87,120 @@ def file_printer(filename):
         if tag == "cmap":
             _ = [st.cmap for st in table.tables]
-        ttfont[tag] = table
 
+    _assert_all_keys_loaded(font)
+
+    # post has a cool pattern of raising errors if it has no glyphOrder attribute
+    # font isn't all that useful w/o glyph order so lets confirm we're ok on that front
+    # TODO contemplate whether this is simply a FontTools bug
+    font.getGlyphOrder()
+
+    return font
+
+
+class OtDataPathEntry(NamedTuple):
+    value: Any
+    attr_name: str  # attr of parent to get value
+
+
+OtDataPath = Tuple[OtDataPathEntry, ...]
+
+
+def traverse_ot_data(root, callback: Callable[[OtDataPath], None]) -> Iterable[Any]:
+    visited = set()
+    frontier: List[OtDataPath] = [(OtDataPathEntry(root, ""),)]
+    while frontier:
+        # path is (value, attr_name) tuples. attr_name is attr of parent to get value
+        path = frontier.pop(0)
+        current = path[-1].value
+        visited.add(id(current))
 
-def reorder_glyphs(ttfont: ttLib.TTFont, glyph_order: List[str]):
+        callback(path)
+
+        converters = getattr(current, "converters", [])
+        for converter in converters:
+            is_simple = isinstance(converter, otConverters.SimpleValue)
+            print("  ", getattr(converter, "name", "?nameless?"), converter, is_simple)
+
+            attr_name = getattr(converter, "name", None)
+            if attr_name:
+                raw_value = getattr(current, attr_name, None)
+            else:
+                print("  ", "^has no name")
+                raw_value = None
+
+            if not isinstance(raw_value, list):
+                raw_value = [raw_value]
+
+            for value in raw_value:
+                if (
+                    value is not None
+                    and hasattr(value, "converters")
+                    and id(value) not in visited
+                ):
+                    frontier.append(path + (OtDataPathEntry(value, attr_name),))
+
+
+def reorder_glyphs(font: ttLib.TTFont, glyph_order: List[str]):
     # Changing the order of glyphs in a TTFont requires that all tables that use
     # glyph indexes have been fully decompiled (loaded with lazy=False).
     # Cf. https://github.com/fonttools/fonttools/issues/2060
-    ensure_ttfont_fully_decompiled(ttfont)
-    ttfont.setGlyphOrder(glyph_order)
+
+    _assert_all_keys_loaded(font)
+
+    font.setGlyphOrder(glyph_order)
     # glyf table is special and needs its own glyphOrder...
-    assert ttfont.isLoaded("glyf")
-    ttfont["glyf"].glyphOrder = glyph_order
+    font["glyf"].glyphOrder = glyph_order
+
+    # make sure reverse glyph order rebuilds too
+    font.getReverseGlyphMap(rebuild=True)
+
+    # TEMPORARY HOPEFULLY
+    # The glyphs in Coverage observably do not update properly, following logs:
+    # otTables.py:581] GSUB/GPOS Coverage is not sorted by glyph ids.
+    # Despite Coverage attempting self-repair resulting font will fail OTS.
+
+    # TODO also GSUB?
```

Comment:
Yes, all of GPOS, GSUB, GDEF and MATH can contain Coverages in various places

Reply:
QQ, should we just delete the current self-repair stuff? - it's very confusing that it looks like it fixes itself but ... doesn't.

Reply:
(and given your clarification of things sorted in parallel I'm not sure it knows enough to ever do it correctly)

Reply:
No rush. I believe your doc explains sufficiently to build at least the outlines of a correct solution we can then iterate on here and hopefully in time promote into FT.